lib/rethinkdb/connection.ex

```diff
@@ -123,213 +119,101 @@ defmodule RethinkDB.Connection do
   end
 
-  def handle_call(_, _,
-    state = %{pending: pending, config: %{max_pending: max_pending}}) when map_size(pending) > max_pending do
-    {:reply, %RethinkDB.Exception.TooManyRequests{}, state}
-  end
-
-  def handle_call({:query_noreply, query}, _from, state = %{token: token}) do
-    new_token = token + 1
-    token = << token :: little-size(64) >>
-    {:noreply, state} = Request.make_request(query, token, :noreply, %{state | token: new_token})
-    {:reply, :noreply, state}
-  end
-
-  def handle_call({:query, query}, from, state = %{token: token}) do
-    new_token = token + 1
-    token = << token :: little-size(64) >>
-    Request.make_request(query, token, from, %{state | token: new_token})
-  end
-
-  def handle_call({:continue, token}, from, state) do
-    query = "[2]"
-    Request.make_request(query, token, from, state)
-  end
+  #
+  # DBConnection API
+  #
 
-  def handle_call({:stop, token}, from, state) do
-    query = "[3]"
-    Request.make_request(query, token, from, state)
+  def connect(options) do
+    # Starts the multiplexer process and stores its pid and some default options to the state.
+    {:ok, pid} = Multiplexer.start_link(options)
+    {:ok, %__MODULE__{pid: pid, options: Keyword.take(options, ~w(db database timeout)a)}}
   end
 
-  def handle_call(:noreply_wait, from, state = %{token: token}) do
-    query = "[4]"
-    new_token = token + 1
-    token = << token :: little-size(64) >>
-    Request.make_request(query, token, from, %{state | token: new_token})
+  def disconnect(_err, _state) do
+    # TODO
   end
 
-  def handle_cast(:stop, state) do
-    {:disconnect, :normal, state};
+  def checkin(state) do
+    {:ok, state}
   end
 
-  def handle_info({proto, _port, data}, state = %{socket: socket}) when proto in [:tcp, :ssl] do
-    :ok = Transport.setopts(socket, [active: :once])
-    Request.handle_recv(data, state)
+  def checkout(state) do
+    {:ok, state}
   end
 
-  def handle_info({closed_msg, _port}, state) when closed_msg in [:ssl_closed, :tcp_closed] do
-    {:disconnect, closed_msg, state}
-  end
+  def handle_execute(%Q{message: message} = query, _params, options, %__MODULE__{pid: pid} = state) do
+    # Overrides default options with specified ones.
+    options = Keyword.merge(state.options, options)
 
-  def handle_info(msg, state) do
-    Logger.debug("Received unhandled info: #{inspect(msg)} with state #{inspect state}")
-    {:noreply, state}
-  end
-
-  def terminate(_reason, %{socket: socket}) do
-    Transport.close(socket)
-    :ok
-  end
+    message =
+      unless message do
+        # This only applies when called with run/3 and is used
+        # to support default options when encoding the query.
+        message = describe(query, options)
+        |> Map.fetch!(:message)
+      else
+        message
+      end
 
-  def terminate(_reason, _state) do
-    :ok
+    case Multiplexer.send_recv(pid, message, options) do
```

Comment:
The [`handle_execute/4`](https://hexdocs.pm/db_connection/DBConnection.html#c:handle_execute/4) callback is  called in the client process. Which means that the connection can handle concurrent queries from multiple processes.

I wrote a small test:
https://gist.github.com/almightycouch/6fed5356c60a6ef70a41eea31777fa62

When i log the output inside `send_recv/4` with following line:
```elixir
IO.inspect "send_recv from #{inspect self} to #{inspect pid}"
```

this is my output:

```
"send_recv from #PID<0.178.0> to #PID<0.176.0>"
"send_recv from #PID<0.177.0> to #PID<0.176.0>"
"send_recv from #PID<0.179.0> to #PID<0.176.0>"
```

Using`:poolboy`:

```elixir
{:ok, pid} = RethinkDB.Connection.start_link, pool: DBConnection.Poolboy
```

this is my output:

```
"send_recv from #PID<0.292.0> to #PID<0.247.0>"
"send_recv from #PID<0.290.0> to #PID<0.251.0>"
"send_recv from #PID<0.291.0> to #PID<0.249.0>"
```

Which is i think what is excepted.